File: framework/languages/java/e2e/PubSub/Server/Subscriber/src/main/java/systems/zlink/e2e/pubsub/subscriber/Infrastructure/EvidenceStore.java
package systems.zlink.e2e.pubsub.subscriber.Infrastructure;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import systems.zlink.e2e.pubsub.shared.Contracts;
import systems.zlink.e2e.pubsub.subscriber.Configuration.SubscriberOptions;

public final class EvidenceStore {
    private final SubscriberOptions options;
    private final List<Contracts.EvidenceEntry> entries = new ArrayList<>();

    public EvidenceStore(SubscriberOptions options) {
        this.options = options;
    }

    public String subscriberRid() {
        return options.rid();
    }

    public boolean accepts(String topic) {
        return options.topics().contains(topic);
    }

    public void delayIfConfigured(String scenario) {
        if (options.delay().delayMillis() <= 0 || !"ps-b1".equals(scenario)) {
            return;
        }
        try {
            Thread.sleep(options.delay().delayMillis());
        } catch (InterruptedException error) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("subscriber delay interrupted", error);
        }
    }

    public synchronized void record(
        String marker,
        String topic,
        String scenario,
        int sequence,
        String value) {
        entries.add(new Contracts.EvidenceEntry(
            marker,
            options.rid(),
            topic,
            scenario,
            sequence,
            value));
    }

    public synchronized Contracts.EvidenceSnapshot snapshot() {
        return new Contracts.EvidenceSnapshot(options.rid(), List.copyOf(entries));
    }

    public Set<String> topics() {
        return options.topics();
    }
}

File: framework/languages/java/e2e-kotlin/PubSub/Server/Subscriber/src/main/kotlin/systems/zlink/e2e/kotlin/pubsub/subscriber/Infrastructure/EvidenceStore.kt
package systems.zlink.e2e.kotlin.pubsub.subscriber

import systems.zlink.e2e.kotlin.pubsub.shared.EvidenceEntry
import systems.zlink.e2e.kotlin.pubsub.shared.EvidenceSnapshot

class EvidenceStore(
    val subscriberRid: String,
    private val topics: Set<String>,
    private val handlerDelayMillis: Long?,
) {
    private val entries = mutableListOf<EvidenceEntry>()

    fun accepts(topic: String): Boolean = topics.contains(topic)

    fun delayIfConfigured(scenario: String) {
        val delay = handlerDelayMillis
        if (delay == null || scenario != "ps-b1") {
            return
        }
        try {
            Thread.sleep(delay)
        } catch (error: InterruptedException) {
            Thread.currentThread().interrupt()
            throw IllegalStateException("subscriber delay interrupted", error)
        }
    }

    @Synchronized
    fun record(
        marker: String,
        topic: String,
        scenario: String,
        sequence: Int,
        value: String,
    ) {
        entries += EvidenceEntry(marker, subscriberRid, topic, scenario, sequence, value)
    }

    @Synchronized
    fun snapshot(): EvidenceSnapshot =
        EvidenceSnapshot(subscriberRid, entries.toList())
}
